pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -619,14 +627,18 @@ func (m *ManagerImpl) readCheckpoint() error {
 
 // UpdateAllocatedDevices frees any Devices that are bound to terminated pods.
 func (m *ManagerImpl) UpdateAllocatedDevices() {
-	activePods := m.activePods()
+	activeAndAdmittedPods := m.activePods()
+	for _, pod := range m.admittedPods {
+		activeAndAdmittedPods = append(activeAndAdmittedPods, pod)
+	}
```

Comment:
here we don't need locks because this function is always called synchronously in the allocation flow, right?

Reply:
...except for podresources API servers, so we need locks here as well.

Reply:
just move the 4 lines above inside the lock; the `if !m.sourcesReady.AllReady()` line should stay like this I believe.
AFAIK/AFAICT, querying for the active pods after the `AllReady` check should have no negative effect.

Reply:
SO I think we can return the stale data under calls to the CPU and memory manager.

Reply:
We do not call the `RemoveContainer` under the post stop internal hook, I think it is still used somewhere internally, but in general it redundant method now. We stopped running it to prevent the problem with container restarts - https://github.com/kubernetes/kubernetes/pull/90377/files

Reply:
It is still questionable why do we need to call `UpdateAllocatedDevices`, I can see we get pods here https://github.com/kubernetes/kubernetes/blob/9ff3b7e744b34c099c1405d9add192adbef0b6b1/pkg/kubelet/apis/podresources/server_v1.go#L54, so we do not really fetch whole pod state information but we do it for a specific set of pods.

Reply:
How do we ensure a (reasonnably) consistent view of device allocations without calling `UpdateAllocatedDevices`?

Reply:
Why the pod allocated devices should be different from one provided during the admission phase?

Reply:
Because `GetPods` returns *all* the pods (it's https://github.com/kubernetes/kubernetes/blob/9ff3b7e744b34c099c1405d9add192adbef0b6b1/pkg/kubelet/pod/pod_manager.go#L260) - the real loop which takes into account the active pods and the actual allocation, like excluding the terminate pods, happens in `UpdateAllocatedDevices`. This code is quite old - it is inherited from the server_v1alpha1.go and we can probably make it smarter taking into account the things we learned in between, but would need a separate PR and some testing. I'm not sure we would need to be bug-compatible with `v1alpha1` API or not.

Reply:
I see, I was sure it should return only active pods. Thanks for the info.

Reply:
Well, you're right we can improve this flow, like a lot. But just not quickly enough to make this PR simpler, unfortunately :\